File: 4.8-bc/ts/stage-3-hexagonal/src/shared/apperr.ts
// shared/apperr — типізовані помилки для крос-BC помилкової семантики.
// Кожен BC мапить власні sentinel errors на AppError через свій ports-шар.

export class AppError extends Error {
  readonly code: string;
  readonly statusCode: number;

  constructor(code: string, message: string, statusCode: number) {
    super(message);
    this.name = "AppError";
    this.code = code;
    this.statusCode = statusCode;
  }
}

export function isAppError(e: unknown): e is AppError {
  return e instanceof AppError;
}

File: 4.8-bc/py/stage-3-hexagonal/app/shared/apperr.py
"""apperr — typed errors for cross-BC error semantics.

Each BC maps its domain sentinel errors to ``AppError`` in its own
``infra/http/errors.py``. ``shared`` knows nothing about specific BCs.
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass
class AppError(Exception):
    code: str
    message: str
    status_code: int

    def __str__(self) -> str:  # pragma: no cover
        return f"{self.code}: {self.message}"
